### PR TITLE
Non-recursive DFS token tree revoke

### LINKED ISSUE #2348: panic: Stack Overflow in revokeTreeSalted
Vault panics when revoking some accessors. Seems to be in the code path of `revokeTreeSalted`

Vault Version: 0.6.3

```
Feb 07 09:58:32 vault01 vault[429]: runtime: goroutine stack exceeds 1000000000-byte limit
Feb 07 09:58:32 vault01 vault[429]: fatal error: stack overflow
Feb 07 09:58:32 vault01 vault[429]: runtime stack:
Feb 07 09:58:32 vault01 vault[429]: runtime.throw(0x16af3c8, 0xe)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/runtime/panic.go:566 +0x95
Feb 07 09:58:32 vault01 vault[429]: runtime.newstack()
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/runtime/stack.go:1061 +0x416
Feb 07 09:58:32 vault01 vault[429]: runtime.morestack()
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/runtime/asm_amd64.s:366 +0x7f
Feb 07 09:58:32 vault01 vault[429]: goroutine 132 [running]:
Feb 07 09:58:32 vault01 vault[429]: runtime.strhash(0xc4ae186430, 0x7a19b95f, 0xc4ae1862f0)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/runtime/alg.go:96 fp=0xc4ae1862c0 sp=0xc4ae1862b8
Feb 07 09:58:32 vault01 vault[429]: type..hash.net/http.connectMethodKey(0xc4ae186430, 0x7a19b95f, 0xc4ae186330)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/client.go:1 +0x35 fp=0xc4ae1862e8 sp=0xc4ae1862c0
Feb 07 09:58:32 vault01 vault[429]: runtime.mapaccess2(0x1472780, 0xc420276090, 0xc4ae186430, 0xc4204eef49, 0x4)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/runtime/hashmap.go:343 +0x73 fp=0xc4ae186330 sp=0xc4ae1862e8
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Transport).getIdleConn(0xc42016c4b0, 0x0, 0xc4204eef49, 0x4, 0xc4204eeef1, 0xe, 0x0, 0x0, 0x0, 0x0)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/transport.go:722 +0x11d fp=0xc4ae186470 sp=0xc4ae186330
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Transport).getConn(0xc42016c4b0, 0xc42ad3b920, 0x0, 0xc4204eef49, 0x4, 0xc4204eeef1, 0xe, 0x0, 0x0, 0xc43981e0e0)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/transport.go:849 +0x11f fp=0xc4ae1867b8 sp=0xc4ae186470
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Transport).RoundTrip(0xc42016c4b0, 0xc42aaa53b0, 0xc42016c4b0, 0x0, 0x0)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/transport.go:367 +0x307 fp=0xc4ae186a10 sp=0xc4ae1867b8
Feb 07 09:58:32 vault01 vault[429]: net/http.send(0xc42aaa53b0, 0x209c660, 0xc42016c4b0, 0x0, 0x0, 0x0, 0x8, 0xc4ae186b90, 0xc4294d5d20)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/client.go:256 +0x15f fp=0xc4ae186b28 sp=0xc4ae186a10
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Client).send(0xc4205411a0, 0xc42aaa53b0, 0x0, 0x0, 0x0, 0xc4294d5d20, 0x0, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/client.go:146 +0x102 fp=0xc4ae186ba0 sp=0xc4ae186b28
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Client).doFollowingRedirects(0xc4205411a0, 0xc42aaa53b0, 0x17b3938, 0x3, 0x58999a01, 0x36411972)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/client.go:528 +0x5e5 fp=0xc4ae186cf0 sp=0xc4ae186ba0
Feb 07 09:58:32 vault01 vault[429]: net/http.(*Client).Do(0xc4205411a0, 0xc42aaa53b0, 0x20e1880, 0x0, 0x2)
Feb 07 09:58:32 vault01 vault[429]:         /goroot/src/net/http/client.go:184 +0x1ea fp=0xc4ae186d50 sp=0xc4ae186cf0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api.(*Client).doRequest(0xc420163920, 0xc42a653360, 0xc4ae186e58, 0xc4ae
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api/api.go:497 +0x8e fp=0xc4ae186dd0 sp=0xc4ae186d50
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api.(*KV).getInternal(0xc4201564f8, 0xc517ba3780, 0x40, 0xc42a4d0e70, 0x
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api/kv.go:164 +0x269 fp=0xc4ae186ef0 sp=0xc4ae186dd0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api.(*KV).Keys(0xc4201564f8, 0xc517ba3780, 0x40, 0x169fd8e, 0x1, 0x0, 0x
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/consul/api/kv.go:142 +0x156 fp=0xc4ae186f90 sp=0xc4ae186ef0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/physical.(*ConsulBackend).List(0xc42008a370, 0xc517ba3700, 0x3a, 0x0, 0x0, 0x0, 0x0, 0x0)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/physical/consul.go:329 +0x290 fp=0xc4ae1870a8 sp=0xc4ae186f90
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/physical.(*Cache).List(0xc4201538a0, 0xc517ba3700, 0x3a, 0x2, 0x2, 0xed02b9148, 0x3640eede, 0x20e1880)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/physical/cache.go:91 +0x47 fp=0xc4ae1870f8 sp=0xc4ae1870a8
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*AESGCMBarrier).List(0xc420163980, 0xc517ba3700, 0x3a, 0x0, 0x0, 0x0, 0x0, 0x0)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/barrier_aes_gcm.go:678 +0x1d8 fp=0xc4ae187170 sp=0xc4ae1870f8
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*BarrierView).List(0xc420289120, 0xc42a4d0e40, 0x30, 0xc43981e120, 0x28, 0x169fd8e, 0x1, 0xc42a4d0e4
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/barrier_view.go:44 +0xb8 fp=0xc4ae1871c0 sp=0xc4ae187170
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43981e120, 0x28, 0xc517ba3280, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:941 +0xa9 fp=0xc4ae187250 sp=0xc4ae1871c0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43effec10, 0x28, 0xc4a39d1940, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae1872e0 sp=0xc4ae187250
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438621bd0, 0x28, 0xc48ea8b3c0, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187370 sp=0xc4ae1872e0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43effe6d0, 0x28, 0xc461297140, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187400 sp=0xc4ae187370
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43effe510, 0x28, 0xc4d0f71840, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187490 sp=0xc4ae187400
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438621310, 0x28, 0xc4d7557080, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187520 sp=0xc4ae187490
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438621150, 0x28, 0xc48c6abb80, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae1875b0 sp=0xc4ae187520
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438620dd0, 0x28, 0xc484463040, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187640 sp=0xc4ae1875b0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438620a50, 0x28, 0xc46b64aec0, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae1876d0 sp=0xc4ae187640
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43a17b620, 0x28, 0xc51a62c540, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187760 sp=0xc4ae1876d0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc4386206d0, 0x28, 0xc4d6c86200, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae1877f0 sp=0xc4ae187760
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43a17b0e0, 0x28, 0xc487aedec0, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187880 sp=0xc4ae1877f0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc438620190, 0x28, 0xc494102a40, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187910 sp=0xc4ae187880
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43a17aba0, 0x28, 0xc4921af140, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae1879a0 sp=0xc4ae187910
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc43a17a9e0, 0x28, 0xc46728c9c0, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187a30 sp=0xc4ae1879a0
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc439ee3540, 0x28, 0xc49b39c340, 0x1)
Feb 07 09:58:32 vault01 vault[429]:         /gopath/src/github.com/hashicorp/vault/vault/token_store.go:950 +0xfe fp=0xc4ae187ac0 sp=0xc4ae187a30
Feb 07 09:58:32 vault01 vault[429]: github.com/hashicorp/vault/vault.(*TokenStore).revokeTreeSalted(0xc420132e40, 0xc439ee3380, 0x28, 0xc46105f280, 0x1)
```

### PR DESCRIPTION
**Pull request by lemondrank**
fixes #2348

### CODE REVIEW

File: vault/token_store.go [hunk 1]
```diff
@@ -1023,29 +1023,40 @@ func (ts *TokenStore) RevokeTree(id string) error {
...
+	dfs := make([]string, 0, 64)
+	dfs = append(dfs, saltedId)
+
+	for l := len(dfs); l > 0; l = len(dfs) {
+		id := dfs[0]
+		path := parentPrefix + id + "/"
+		children, err := ts.view.List(path)
+		if err != nil {
+		  return fmt.Errorf("failed to scan for children: %v", err)
+    }
```
Comment (briankassouf) [inline vault/token_store.go]: Could you run go fmt on your code? https://golang.org/cmd/gofmt/

File: vault/token_store.go [hunk 2]
```diff
@@ -1023,29 +1023,40 @@ func (ts *TokenStore) RevokeTree(id string) error {
...
+				return fmt.Errorf("failed to revoke entry: %v", err)
+			}
+			// If the length of l is equal to 1, then the last token has been deleted
+			if l == 1 { 
+				return nil
+			} else {
+				dfs = dfs[1:l] 
+			}
+		} else { 
+		  // If we make it here, the there are children and they must
```
Comment (briankassouf) [inline vault/token_store.go]: type on "the there"

File: vault/token_store.go [hunk 3]
```diff
@@ -1120,7 +1131,7 @@ func (ts *TokenStore) handleTidy(req *logical.Request, data *framework.FieldData
 	// Scan through the secondary index entries; if there is an entry
 	// with the token's salt ID at the end, remove it
 	for _, parent := range parentList {
-		children, err := ts.view.List(parentPrefix + parent + "/")
```
Comment (briankassouf) [inline vault/token_store.go]: How come this changed?

File: vault/token_store.go [hunk 4]
```diff
@@ -1032,29 +1032,41 @@ func (ts *TokenStore) RevokeTree(id string) error {
...
+		// If the length of the children array is zero,
+		// then we are at a leaf node.
+		if len(children) == 0 {
+			if err := ts.revokeSalted(id); err != nil {
+				return fmt.Errorf("failed to revoke entry: %v", err)
+			}
+			// If the length of l is equal to 1, then the last token has been deleted
+			if l == 1 {
+				return nil
+			} else {
```
Comment (chrishoffman) [inline vault/token_store.go]: This `else` is not necessary.

File: vault/token_store.go [hunk 5]
```diff
@@ -1032,29 +1032,41 @@ func (ts *TokenStore) RevokeTree(id string) error {
...
-	}
-
-	// Recursively nuke the children. The subtle nuance here is that
-	// we don't have the acutal ID of the child, but we have the salted
-	// value. Turns out, this is good enough!
-	for _, child := range children {
-		if err := ts.revokeTreeSalted(child); err != nil {
-			return err
+	// Initialize slice to be arbitraty capacity of 64
+	dfs := make([]string, 0, 64)
```
Comment (chrishoffman) [inline vault/token_store.go]: I could be wrong but I'm not sure you are buying yourself anything here with allocating the slice size.  A simple `var dfs []slice` should do, I believe.

File: vault/token_store.go [hunk 6]
```diff
@@ -1032,29 +1032,41 @@ func (ts *TokenStore) RevokeTree(id string) error {
...
+		// then we are at a leaf node.
+		if len(children) == 0 {
+			if err := ts.revokeSalted(id); err != nil {
+				return fmt.Errorf("failed to revoke entry: %v", err)
+			}
+			// If the length of l is equal to 1, then the last token has been deleted
+			if l == 1 {
+				return nil
+			} else {
+				dfs = dfs[1:l]
```
Comment (chrishoffman) [inline vault/token_store.go]: You can actually just drops off the `l` which will default to length of the slice.

**Comment by chrishoffman** [comment]
LGTM!

**Comment by lemondrank** [comment]
sweet! thanks for all the help!

**Comment by jefferai** [comment]
I'll try to get to a review soon but adding to 0.7.3.

**Comment by lemondrank** [comment]
sounds great! thanks!

**Pull request merged**